src/sdk/pynni/nni/compression/torch/builtin_pruners.py

```diff
@@ -206,25 +259,141 @@ def _get_distance_sum(self, weight, in_idx, out_idx):
             between this specified filter and all other filters.
         in_idx: int
             input channel index of specified filter
-
         Returns
         -------
         float32
             The total distance
         """
         logger.debug('weight size: %s', weight.size())
-        if len(weight.size()) == 4: # Conv2d
+        if len(weight.size()) == 4:  # Conv2d
             w = weight.view(-1, weight.size(-2), weight.size(-1))
             anchor_w = weight[out_idx, in_idx].unsqueeze(0).expand(w.size(0), w.size(1), w.size(2))
-        elif len(weight.size()) == 3: # Conv1d
+        elif len(weight.size()) == 3:  # Conv1d
             w = weight.view(-1, weight.size(-1))
             anchor_w = weight[out_idx, in_idx].unsqueeze(0).expand(w.size(0), w.size(1))
         else:
             raise RuntimeError('unsupported layer type')
         x = w - anchor_w
-        x = (x*x).sum((-2, -1))
+        x = (x * x).sum((-2, -1))
         x = torch.sqrt(x)
         return x.sum()
 
     def update_epoch(self, epoch):
         self.epoch_pruned_layers = set()
+
+
+class FilterPruner(Pruner):
+    """
+    A structured pruning algorithm that prunes the filters of smallest magnitude
+    weights sum in the convolution layers to achieve a preset level of network sparsity.
+    Hao Li, Asim Kadav, Igor Durdanovic, Hanan Samet and Hans Peter Graf,
+    "PRUNING FILTERS FOR EFFICIENT CONVNETS", 2017 ICLR
+    https://arxiv.org/abs/1608.08710
+    """
+
+    def __init__(self, model, config_list):
+        """
+        Parameters
+        ----------
+        model : torch.nn.module
+            Model to be pruned
+        config_list : list
+            List on pruning configs
+        """
+
+        super().__init__(model, config_list)
+        self.if_init_list = {}
+
+    def calc_mask(self, layer, config):
+        """
+        Calculate the mask of given layer.
+        Filters with the smallest sum of its absolute kernel weights are masked.
+        Parameters
+        ----------
+        layer : LayerInfo
+            the layer to instrument the compression operation
+        config : dict
+            layer's pruning config
+        Returns
+        -------
+        torch.Tensor
+            mask of the layer's weight
+        """
+
+        weight = layer.module.weight.data
+        op_name = layer.name
+        op_type = layer.type
+        assert op_type == 'Conv2d', 'FilterPruner only supports 2d convolution layer pruning'
+        if self.if_init_list.get(op_name, True):
```

Comment:
It is not straightforward when the default value is true,  suggest to refactor code (such as rename if_init_list) to make this condition easier to understand.

Reply:
`self.if_init_list` is used for storing flags and denote if the layer's mask needs to be re-calculated, at the beginning all the layers to be pruned need to re-calculate the mask, so we set the default value is true.

Reply:
Your are right. My suggestion is to make this logic to be easier understood by others.
Since the value is boolean, it is actually can be replaced by a set.
What if rename self.if_init_list  to `mask_calculated_ops` and make it a set,  then put the op_name into the set after calculating its mask?

Reply:
I've refactored this, thanks